batch_job/src/batch_job_k8s.c

```diff
@@ -0,0 +1,463 @@
+		char *k8s_config_fn = string_format("%s.json", pod_id);
+
+		FILE *fd = fopen(k8s_config_fn, "w+");
+		
+		if(resources) {
+			int cores = resources->cores;
+			int memory = resources->memory;
+			char *k8s_cpu, *k8s_memory;
+			if (cores > -1 && memory > -1) {
+				k8s_cpu = string_format("%dm", cores * 1000);
+				k8s_memory = string_format("%dMi", memory);
+			} else {
+				// By default, each container will request
+				// 0.5 cpu and 1024 MB ram
+				k8s_cpu = string_format("%dm", 500);
+				k8s_memory = string_format("%dMi", 1024);
+			}
+			resources_block = string_format(resource_tmpl, 
+					k8s_cpu, k8s_memory, k8s_cpu, k8s_memory);	
+		}
+
+		fprintf(fd, k8s_config_tmpl, mf_uuid->str, pod_id, pod_id, k8s_image, 
+				resources_block, job_id, pod_id);
+
+		fclose(fd);
+
+		if(access(k8s_script_file_name, F_OK | X_OK) == -1) {
+			debug(D_BATCH, "Generating k8s script...");
+			FILE *f = fopen(k8s_script_file_name, "w");
+			fprintf(f, "%s", k8s_script);
+			fclose(f);
+			// Execute permissions
+			chmod(k8s_script_file_name, 0755);
+		}
+
+		char *job_id_str = string_format("%d", job_id);
+
+		execlp("/bin/bash", "bash", k8s_script_file_name, "create", pod_id, job_id_str, 
+				extra_input_files, cmd, extra_output_files, (char *) NULL);
+		_exit(127);
+
+	} else {
+
+		debug(D_BATCH, "couldn't create new process: %s\n", strerror(errno));
+		return -1;
+
+	}
+
+	return -1;
+}
+
+static int batch_job_k8s_remove (struct batch_queue *q, batch_job_id_t jobid)
+{
+	
+	pid_t pid = fork();
+	char *pod_id = string_format("%s-%d", mf_uuid->str, (int)jobid);
+	
+    if(pid > 0) {
+
+    	int status;
+
+    	debug(D_BATCH, "Trying to remove task %d by deleting pods %s.", 
+			(int)jobid, pod_id);
+		
+		// wait child process to complete
+
+		if(waitpid(pid, &status, 0) == -1) {
+			fatal("Failed to remove pods %s", pod_id);
+		}
+
+		if(WIFEXITED(status)) {
+			debug(D_BATCH, "Successfully delete pods %s", pod_id);
+		}
+		
+		free(pod_id);
+		return 0;
+
+    } else if(pid < 0) {
+
+		fatal("couldn't create new process: %s\n", strerror(errno));
+
+    } else {
+
+		char *cmd = string_format("kubectl delete pods %s", pod_id);
+
+		execlp("/bin/bash", "bash", "-c", cmd, (char *) 0);
+		_exit(errno);
+
+    }
+
+    return 1;	
+
+}
+
+static void batch_job_k8s_handle_complete_task (char *pod_id, int job_id,
+		k8s_job_info *curr_k8s_job_info, int exited_normally, 
+		struct batch_job_info *info_out, struct list *running_pod_lst,
+		struct batch_queue *q)
+{
+	struct batch_job_info *info;
+	int timeout = 5;
+	debug(D_BATCH, "%d is failed to execute.", job_id);
+	info = itable_remove(q->job_table, job_id);
+	info->exited_normally = exited_normally;
+	// TODO get exact exit_code
+	if (exited_normally == 0) {
+		info->exit_code = 1;
+	}
+	memcpy(info_out, info, sizeof(*info));
+	free(info);
+
+	list_remove(running_pod_lst, pod_id);
+	batch_job_k8s_remove(q, job_id);
+
+	if(curr_k8s_job_info->is_running == 0) {
```

Comment:
This seems very strange.
You need to check the return value of process_wait, otherwise you can't tell if anything exited or not.
And the returned structure needs to be deleted.